### PR TITLE
Made auth support generic

### PR DESCRIPTION
**Pull request by mcollina**
@delvedor what do you think?

There are a bunch of TODOs, and it will probably require more testing.

Do you think we can easily use the `'onRequest'` hook to apply this to protect all routes in a plugin?

### CODE REVIEW

**Comment by delvedor** [comment]
I like the idea of the queue to support multiple authentication strategies, but I think is too cryptic for the developer. We should find I way to make it more "comfortable".

**Comment by delvedor** [comment]
Within the latest version of [Fastify](https://github.com/fastify/fastify/releases) we can pass an array of functions to `beforeHandler`, so we can avoid the *auth* field and handle everything inside the *beforeHandler*.

In addition I think could be a good idea that in case of the JWT validation add the `jwt` decorator to `Request`, and save inside the decoded jwt.

```js
fastify.decorateRequest('jwt', null)

...

jwt.verify(request.req.headers['auth'], onVerify)

function onVerify (err, decoded) {
  if (err || !decoded.user || !decoded.password) {
    return done(null, false, 'Token not valid')
 }
  request.jwt = decoded
  ...
}

...

// access the decoded data
console.log(request.jwt.user)
```

**Comment by mcollina** [comment]
The question is.. do we need a fastify-auth at all then?
Maybe we should just support some standard places to put routes for unauthenticated users and such.

**Comment by delvedor** [comment]
I thought about this and I think that the best solution is a little rework of yours:

`beforeHandler` takes an array of functions, we can't pass all the auth functions to `beforeHandler` because if one fails everything will fail, so we need something like the following:
```js
function runAuth (functions) {
  function _runAuth (request, reply, done) {
    var functions = this.functions
    var valid = false
    var i = 0
    nextAuth()

    function nextAuth () {
      var func = functions[i++]

      if (!func) {
        if (valid) {
          done()
        } else {
          if (!reply.res.statusCode || reply.res.statusCode < 400) {
            reply.code(401)
          }
          done(new Error())
        }
        return
      }

      func.call(this, request, reply, onAuth)
    }

    function onAuth (err) {
      if (err) {
        done(err)
        return
      }

      valid = true
      nextAuth()
    }
  }

  return _runAuth.bind({ functions })
}
```
We should also use something like [reusify](https://github.com/mcollina/reusify) to reuse the functions.

With the above code we can run multiple auth strategies, and return an error only if all the stretegies are failing.

Thoughts?

**Comment by mcollina** [comment]
Yes! Would you like to update this proposal?

**Comment by delvedor** [comment]
The cool thing of this solution is that could work in the same way both in the `beforeHandler` and the `preHandler` hook, so you can use it on a plugin level or on a route level 👍

File: example.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+      handler: (req, reply) => {
+        req.log.info('Auth free route')
+        reply.send({ hello: 'world' })
+      }
+    })
+
+    fastify.route({
+      method: 'GET',
+      url: '/auth',
+      beforeHandler: fastify.auth([fastify.verifyJWTandLevel.bind(fastify)]),
```
Comment (mcollina) [inline example.js]: Do we really need the `bind()` call?

Comment (delvedor) [inline example.js]: `auth` does not need it, while `terifyJWTandLevel` needs it.

Comment (mcollina) [inline example.js]: we can avoid it either by doing  `bind` within `fastify.auth`, or by using `func.call` [here](https://github.com/fastify/fastify-auth/pull/2/files#diff-07eb183747dcb27f01bb7310efb53fc8R35)

Comment (delvedor) [inline example.js]: Does not affect performances?

Comment (mcollina) [inline example.js]: doing `bind` beforehand inside `fastify.auth` does not for sure.

File: README.md [hunk 2]
```diff
@@ -2,29 +2,61 @@
...
+  ]),
+  handler: (req, reply) => {
+    req.log.info('Auth route')
+    reply.send({ hello: 'world' })
+  }
+})
 ```
+`fastify-auth` will run all your authentication methods if at least one succeed it will continue your request, otherwise it wil return an error to the client. It will automatically stop as soon as a authentication succeed.
+
+You can use this plugin on route level, as in the above example or on hook level, by using the `preHandler` hook:
```
Comment (mcollina) [inline README.md]: Can you rephrase a bit, what does 'hook level' means? Is it a plugin-level, via a hook?
